akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala

```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
   implicit val ec: ExecutionContextExecutor = context.dispatcher
   override def supervisorStrategy: SupervisorStrategy = SupervisorStrategy.stoppingStrategy
 
+  private var snapshotCollectors: Set[ActorRef] = Set.empty
+
   def receive: Receive = {
     case Materialize(props, name) =>
       val impl = context.actorOf(props, name)
       sender() ! impl
     case GetChildren =>
       sender() ! Children(context.children.toSet)
-    case GetChildrenSnapshots =>
-      takeSnapshotsOfChildren().map(ChildrenSnapshots.apply _).pipeTo(sender())
-
+    case GetChildrenSnapshots(timeout) =>
+      val collector =
+        context.actorOf(
+          SnapshotCollector
+            .props(context.children.toSet -- snapshotCollectors, timeout, sender())
+            .withDispatcher(context.props.dispatcher))
+      context.watchWith(collector, CollectorCompleted(collector))
+      snapshotCollectors += collector
     case StopChildren =>
       context.children.foreach(context.stop)
       sender() ! StoppedChildren
+    case CollectorCompleted(collector) =>
+      snapshotCollectors -= collector
   }
 
-  def takeSnapshotsOfChildren(): Future[immutable.Seq[StreamSnapshot]] = {
-    // Arbitrary timeout but should always be quick, the failure scenario is that
-    // the child/stream stopped, and we do retry below
-    implicit val timeout: Timeout = 1.second
-    def takeSnapshot() = {
-      val futureSnapshots =
-        context.children.toList.map(child => (child ? ActorGraphInterpreter.Snapshot).mapTo[StreamSnapshot])
-      Future.sequence(futureSnapshots)
+  override def postStop(): Unit = haveShutDown.set(true)
+}
+
+@InternalApi
+private[akka] object SnapshotCollector {
+  case object SnapshotTimeout
+  def props(streamActors: Set[ActorRef], timeout: FiniteDuration, replyTo: ActorRef): Props =
+    Props(new SnapshotCollector(streamActors, timeout, replyTo))
+}
+
+@InternalApi
+private[akka] final class SnapshotCollector(streamActors: Set[ActorRef], timeout: FiniteDuration, replyTo: ActorRef)
+    extends Actor
+    with Timers {
+
+  import SnapshotCollector._
+
+  var leftToRespond = streamActors
+  var collected: List[StreamSnapshot] = Nil
+  if (streamActors.isEmpty) {
+    replyTo ! StatusReply.Success(StreamSupervisor.ChildrenSnapshots(Nil))
+    context.stop(self)
+  } else {
+    streamActors.foreach { ref =>
+      context.watch(ref)
+      ref ! ActorGraphInterpreter.Snapshot
     }
+  }
 
-    // If the timeout hits it is likely because one of the streams stopped between looking at the list
-    // of children and asking it for a snapshot. We retry the entire snapshot in that case
-    retry(() => takeSnapshot(), 3)
+  timers.startSingleTimer(Timeout, Timeout, timeout)
+
+  override def receive: Receive = {
+    case snap: StreamSnapshot =>
+      collected = snap :: collected
+      leftToRespond -= sender()
+      completeIfDone()
+    case Terminated(streamActor) =>
+      leftToRespond -= streamActor
+      completeIfDone()
```

Comment:
If a stream stopped between supervisor extracting the list and getting a snapshot, we just ignore that child (I think that makes most sense)